File: backend/app/schemas/billing.py
"""Greena — Billing schemas (API v1).

The client names *which* plan (by id); it never supplies an amount. The price is
always derived server-side from ``subscription_plans``.
"""

from uuid import UUID

from app.schemas.base import AGRIOSSchema


class InitializePaymentIn(AGRIOSSchema):
    """Request body for POST /billing/initialize."""

    organization_id: UUID
    plan_id: UUID
    callback_url: str | None = None


class InitializePaymentOut(AGRIOSSchema):
    """Where to send the customer to pay, plus the server-derived amount."""

    authorization_url: str
    reference: str
    amount_kes: int
    plan_id: UUID
    plan_name: str


class PlanOut(AGRIOSSchema):
    """A subscription plan for the checkout UI. Prices come from the DB.

    ``is_self_serve`` = a real positive price the checkout can charge (free is 0,
    a custom/Enterprise plan is -1).
    """

    id: UUID
    name: str
    display_name: str
    price_kes: int
    is_self_serve: bool


class CommercialPolicyOut(AGRIOSSchema):
    """The authoritative commercial policy for every public/app surface.

    One endpoint so the website, signup, dashboard, billing and ARIA all read
    the same trial length and Premium price instead of hardcoding their own.
    """

    trial_days: int
    plan_name: str          # customer-facing name of the trial/premium plan
    monthly_price: int      # premium monthly price
    currency: str           # ISO currency code, e.g. "KES"
    billing_period: str     # "month"


class PublicPlanOut(AGRIOSSchema):
    """A subscription plan for the PUBLIC pricing page (no auth required).

    Contains only non-sensitive, already-public catalogue information — price and
    plan limits — so the marketing site reads the same source of truth the
    checkout charges from and cannot silently drift. ``-1`` means unlimited.
    """

    name: str
    display_name: str
    price_kes: int
    is_self_serve: bool
    max_farms: int
    max_houses_per_farm: int
    max_active_flocks: int
    max_aria_queries_per_month: int
    history_days: int
    max_team_members: int


class PaymentStatusOut(AGRIOSSchema):
    """Result of a webhook/verify: what happened and whether the sub is active."""

    reference: str | None = None
    status: str  # activated | already_processed | ignored
    plan_id: UUID | None = None
    subscription_active: bool = False

File: frontend/src/api/billing.ts
/**
 * Greena — Billing API
 *
 * Paystack subscription checkout. The client only names which plan it wants;
 * the backend derives the amount from the plan (never sent from here).
 */

import apiClient from "./client";
import type { APISuccess } from "@/types";

export interface Plan {
  id: string;
  name: string;
  display_name: string;
  price_kes: number;
  is_self_serve: boolean;
}

export interface InitializePaymentResult {
  authorization_url: string;
  reference: string;
  amount_kes: number;
  plan_id: string;
  plan_name: string;
}

export interface PaymentStatus {
  reference: string | null;
  status: string; // activated | already_processed | ignored
  plan_id: string | null;
  subscription_active: boolean;
}

export interface InitializePaymentInput {
  organization_id: string;
  plan_id: string;
  callback_url?: string;
}

export interface TrialStatus {
  is_trial: boolean;
  active: boolean;
  trial_ends_at: string | null;
  days_remaining: number;
  /** Trial-length policy from the backend (single source of truth). */
  trial_days?: number;
}

export interface ReferralStatus {
  referral_code: string | null;
  has_referral: boolean;
  referrer_org_id: string | null;
  reward_status: string | null;
  entry_open: boolean;
  entry_deadline: string | null;
}

export interface ReferralValidation {
  valid: boolean;
  reason: string | null;
}

export interface CreditEntry {
  amount_kes: number;
  source: string;
  payment_reference: string | null;
  balance_after: number;
  created_at: string;
}

export interface CreditBalance {
  balance_kes: number;
  entries: CreditEntry[];
}

export const billingAPI = {
  listPlans: async (): Promise<Plan[]> => {
    const res = await apiClient.get<APISuccess<Plan[]>>("/billing/plans");
    return res.data.data;
  },

  getTrialStatus: async (orgId: string): Promise<TrialStatus> => {
    const res = await apiClient.get<APISuccess<TrialStatus>>(`/billing/trial/${orgId}`);
    return res.data.data;
  },

  getReferralStatus: async (orgId: string): Promise<ReferralStatus> => {
    const res = await apiClient.get<APISuccess<ReferralStatus>>(`/billing/referral/${orgId}`);
    return res.data.data;
  },

  validateReferral: async (orgId: string, code: string): Promise<ReferralValidation> => {
    const res = await apiClient.get<APISuccess<ReferralValidation>>(
      `/billing/referral/${orgId}/validate`,
      { params: { code } },
    );
    return res.data.data;
  },

  submitReferral: async (orgId: string, code: string): Promise<ReferralStatus> => {
    const res = await apiClient.post<APISuccess<ReferralStatus>>(
      `/billing/referral/${orgId}`,
      { code },
    );
    return res.data.data;
  },

  getCredits: async (orgId: string): Promise<CreditBalance> => {
    const res = await apiClient.get<APISuccess<CreditBalance>>(`/billing/credits/${orgId}`);
    return res.data.data;
  },

  initialize: async (payload: InitializePaymentInput): Promise<InitializePaymentResult> => {
    const res = await apiClient.post<APISuccess<InitializePaymentResult>>(
      "/billing/initialize",
      payload,
    );
    return res.data.data;
  },

  verify: async (reference: string): Promise<PaymentStatus> => {
    const res = await apiClient.get<APISuccess<PaymentStatus>>(
      `/billing/verify/${reference}`,
    );
    return res.data.data;
  },
};
